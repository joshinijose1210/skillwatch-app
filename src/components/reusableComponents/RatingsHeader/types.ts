export interface RatingBadge {
    rating: number;
    subTitle: string;
    titleColor: string;
    bgColor: string;
}

export type Props = {
    title: string;
    ratingList: RatingBadge[];
    reviewCycle?: string | number;
    isLoading: boolean;
};
