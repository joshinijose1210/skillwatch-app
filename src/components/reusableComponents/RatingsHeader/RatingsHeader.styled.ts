import { Text } from '@medly-components/core';
import styled from 'styled-components';
import { BadgeWrapper as defaultBadge } from '../Badge/Badge.styled';

export const StyledBadge = styled(defaultBadge)`
    display: flex;
    position: relative;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: 1rem;
    border-radius: 0.5rem;
    width: 22rem;
    height: 12rem;
    padding: 0;
    flex: 1 1 0;
    cursor: pointer;
    transition: transform 0.3s;

    &::after {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
        transition: opacity 2s cubic-bezier(0.165, 0.84, 0.44, 1);
        box-shadow: 0 8px 17px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.15);
        content: '';
        opacity: 0;
        z-index: -1;
    }

    &:hover,
    &:focus {
        transform: scale3d(1.006, 1.006, 1);

        &::after {
            opacity: 1;
        }
    }
`;

export const BadgeWrapper = styled('div')`
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    gap: 2rem;
`;

export const Container = styled('div')`
    display: flex;
    flex-direction: column;
    gap: 1rem;
`;

export const TileTitle = styled(Text)`
    color: ${({ theme }) => theme.pageTitleColor};
    font-weight: 600;
`;

export const StyledRatingCardText = styled(Text).attrs({
    textVariant: 'h4',
    textWeight: 'Regular'
})`
    font-size: ${({ theme }) => theme.analyticsFontSize.content};
`;
