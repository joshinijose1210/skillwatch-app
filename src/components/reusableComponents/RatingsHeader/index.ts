export { RatingsHeader as default } from './RatingsHeader';
