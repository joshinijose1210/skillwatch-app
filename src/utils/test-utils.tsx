import { ToastContainer } from '@medly-components/core';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { store as rtkStore, setupStore } from '@slice';
import { RenderOptions, RenderResult, render } from '@testing-library/react';
import { defaultTheme } from '@theme';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import reduxMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { userMockState } from '../mocks/userMockState';
import { initialState } from '../slice';

interface CustomOptions extends RenderOptions {
    initialState?: any;
    initialEntries?: any;
}

interface WithThemeProviderProps {
    children?: React.ReactNode;
}

const mockStore = reduxMockStore(),
    store = mockStore({
        ...initialState,
        user: { ...initialState.user, groups: ['admin'] }
    });

const WithThemeProvider: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <ThemeProvider theme={defaultTheme}>
        <>
            <ToastContainer position="top-end" />
            {children}
        </>
    </ThemeProvider>
);

const WithStore: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <WithThemeProvider>{children}</WithThemeProvider>
    </Provider>
);

const WithRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <MemoryRouter>
        <WithThemeProvider>{children}</WithThemeProvider>
    </MemoryRouter>
);

const WithStoreAndRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <MemoryRouter>
            <WithThemeProvider>{children}</WithThemeProvider>
        </MemoryRouter>
    </Provider>
);

export const RTKRender = (ui: React.ReactElement, { initialState = userMockState, ...renderOptions }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);

    const RtkWrapper: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
        <Provider store={store}>
            <MemoryRouter initialEntries={renderOptions.initialEntries}>
                <WithThemeProvider>{children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
    return { rtkStore, ...render(ui, { wrapper: RtkWrapper, ...renderOptions }) };
};
const customRender = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithThemeProvider, ...options });

export const renderWithStore = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStore, ...options });

export const renderWithRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithRouter, ...options });

export const renderWithStoreAndRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStoreAndRouter, ...options });

export const wrapper = (props: { children: React.ReactNode }, { initialState = userMockState }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);
    return (
        <Provider store={store}>
            <MemoryRouter>
                <WithThemeProvider>{props.children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render, store };
