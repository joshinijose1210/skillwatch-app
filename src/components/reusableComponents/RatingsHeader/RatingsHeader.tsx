import { routeConstants } from '@constants';
import { Text } from '@medly-components/core';
import { DotsLoader } from '@medly-components/loaders';
import { FC, memo, useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { BadgeWrapper, Container, StyledBadge, StyledRatingCardText, TileTitle } from './RatingsHeader.styled';
import { Props } from './types';

export const RatingsHeader: FC<Props> = memo(({ title, ratingList, isLoading, reviewCycle }) => {
    const navigate = useNavigate();
    const handleNavigation = useCallback(
        (title: string) => {
            navigate(routeConstants.ratings, {
                state: { title: `${title.split(' ')[0].toLowerCase()}${title.split(' ')?.[1] ?? ''}`, reviewCycle, ratingLabel: title }
            });
        },
        [navigate, reviewCycle]
    );
    return (
        <Container>
            <TileTitle textVariant="h4">{title}</TileTitle>
            <BadgeWrapper>
                {ratingList.map((ratingItem, index) => (
                    <StyledBadge
                        key={`${ratingItem.subTitle}_${index}`}
                        bgColor={ratingItem.bgColor}
                        onClick={() => handleNavigation(ratingItem.subTitle)}
                    >
                        {isLoading ? (
                            <DotsLoader />
                        ) : (
                            <Text textVariant="h2" textWeight="Strong" textColor={ratingItem.titleColor}>
                                {ratingItem.rating}
                            </Text>
                        )}
                        <StyledRatingCardText>{ratingItem.subTitle}</StyledRatingCardText>
                    </StyledBadge>
                ))}
            </BadgeWrapper>
        </Container>
    );
});

RatingsHeader.displayName = 'RatingsHeader';
