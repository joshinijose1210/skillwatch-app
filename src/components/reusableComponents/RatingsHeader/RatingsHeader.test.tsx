import { RTKRender } from '@utils/test-utils';
import { RatingsHeader } from './RatingsHeader';

describe('RatingsHeader', () => {
    it('should render properly', () => {
        const { container } = RTKRender(
            <RatingsHeader
                title="dummy Title"
                ratingList={[{ rating: 5, subTitle: 'dummy subtitle', titleColor: '#BC1435', bgColor: '#F64D6E26' }]}
                isLoading={false}
            />
        );
        expect(container).toMatchSnapshot();
    });
});
