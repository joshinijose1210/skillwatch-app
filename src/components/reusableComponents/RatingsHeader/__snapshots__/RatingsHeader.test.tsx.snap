// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingsHeader should render properly 1`] = `
<div>
  <div
    class="sc-1t1ahfa-2 ebUScI"
  >
    <h4
      class="sc-1ntgl3l-0 iEGCWU sc-1t1ahfa-3 gCnYIy"
      textalign="initial"
      textcolor="inherit"
      textvariant="h4"
    >
      dummy Title
    </h4>
    <div
      class="sc-1t1ahfa-1 jOrojb"
    >
      <div
        bgcolor="#F64D6E26"
        class="sc-1exx2co-0 sc-1t1ahfa-0 TPGAq cgZUsL"
      >
        <h2
          class="sc-1ntgl3l-0 TyhST"
          textalign="initial"
          textcolor="#BC1435"
          textvariant="h2"
          textweight="Strong"
        >
          5
        </h2>
        <h4
          class="sc-1ntgl3l-0 fgmHgf sc-1t1ahfa-4 jrLGQQ"
          textalign="initial"
          textcolor="inherit"
          textvariant="h4"
          textweight="Regular"
        >
          dummy subtitle
        </h4>
      </div>
    </div>
  </div>
</div>
`;
